### PR TITLE
Fixed DNS resolution in `OutgoingSelector`

### LINKED ISSUE #2283: Outgoing local filter breaks all outgoing connections
### Bug Description

When using an outgoing local filter with a hostname that does not exist locally, all outgoing connections seem to be broken.

Example config:
```json
{
    "feature": {
        "network": {
            "outgoing": {
                "tcp": true,
                "filter": {
                    "local": ["i-do-not-exist-at-all"]
                }
            },
            "dns": true
        }
    }
}

```
Example app:
```rust
use std::net::TcpStream;

fn main() {
    let _stream = TcpStream::connect("py-serv:80").expect("py-serv should exist in the cluster");
}
```

### Steps to Reproduce

1. Deploy the `py-serv` service in the cluster
2. Run the example app and config from bug description
3. `TcpStream::connect` fails with `Os { code: 5, kind: Uncategorized, message: "Input/output error" }`. Layer prints `2024-03-01T10:28:50.776415Z ERROR ThreadId(01) mirrord_layer::error: Error occured in Layer >> IO(Custom { kind: Uncategorized, error: "failed to lookup address information: Temporary failure in name resolution" })`

### Backtrace

_No response_

### Relevant Logs

_No response_

### Your operating system and version

Linux 6.5.0-21-generic #21~22.04.1-Ubuntu

### Local process

---

### Local process version

_No response_

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by Razz4780**
Closes #2283 

When the user app wants to initiate an outgoing connection, all outgoing filters are resolved in order to determine whether the connection should be done remotely or locally. The issue was that all resolution failures were treated as fatal errors (and were aborting the `connect` call). This PR filters out resolution errors which mean only that the name does not map to any known IP address (which is the case when the user puts a non-existent host in the filter).

Also, this PR adds a `GetAddrInfoResponse` cache on the internal proxy level. No need to do multiple dns queries each time an outgoing connection is initiated

### CODE REVIEW

File: mirrord/intproxy/src/proxies/simple.rs [hunk 1]
```diff
@@ -187,3 +206,43 @@ impl BackgroundTask for SimpleProxy {
...
+    /// [`GetAddrInfoRequest`] this struct knows of.
+    ///
+    /// # Panic
+    ///
+    /// This method panics if this struct does not now of any incomplete [`GetAddrInfoRequest`].
+    fn response_came(&mut self, lookup: DnsLookup) {
+        let host = self
+            .outstanding_requests
+            .pop_front()
+            .expect("received too many GetAddrInfoResponses");
```
Comment (meowjesty) [inline mirrord/intproxy/src/proxies/simple.rs:239]: Does this ever happen?

Wouldn't it be better if this was just a noop in this case? (maybe also adding a log so we could debug it)

Something like:

```rust
fn foo() -> Option<()> {
   let host = foo.pop_front()?;
    ...

    Some(()) // and we never check this anywhere
}
```

File: changelog.d/2283.fixed.md [hunk 2]
```diff
@@ -0,0 +1 @@
+Fixed a bug where non-existent hosts in outgoing filter would prevent the application from making initiating outgoing connections.
```
Comment (meowjesty) [inline changelog.d/2283.fixed.md:1]: > from making initiating outgoing connections

"making initiating" sounds a bit weird? Maybe it's just me though.

**Review by meowjesty — CHANGES_REQUESTED** [review]
YADI®

File: changelog.d/2283.fixed.md (hunk 2 above)
Comment (Razz4780) [reply changelog.d/2283.fixed.md:1]: Ah sorry, should be just "initiating"

File: mirrord/intproxy/src/proxies/simple.rs (hunk 1 above)
Comment (Razz4780) [reply mirrord/intproxy/src/proxies/simple.rs:239]: This can literally never happen :x

Comment (meowjesty) [reply mirrord/intproxy/src/proxies/simple.rs:239]: Not with that attitude!

**Pull request merged**